### ISSUE TITLE
`l2_normalize` has wrong gradient when input is zero

### ISSUE THREAD

**Issue by VictoriaGriffith** [tensorflow/tensorflow#56796]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.9

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.9

### Current Behaviour?

```shell
When the input of `tf.nn.l2_normalize` is a zero tensor, `tf.GradientTape` outputs a very large gradient `1000000`.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
x = tf.random.uniform([1,1], maxval=0,dtype=tf.float64)
with tf.GradientTape(persistent=True,) as g:
  g.watch(x)
  y = tf.nn.l2_normalize(x)
print("gradient", g.gradient(y, x))
```

### Relevant log output

```shell
gradient tf.Tensor([[1000000.]], shape=(1, 1), dtype=float64)
```
</details>

**Comment by VictoriaGriffith** [comment]
Hi @tilakrayal , the bug only exists when the input is all `0`, so if you change the `maxval` to `0` in [gist](https://colab.research.google.com/gist/tilakrayal/61b6c09b97a8f9a23aa2f8941cbcf982/untitled433.ipynb) you can observe the error

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.8, v2.9 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/adb20c3861295b14d97a3e8f23e90885/56796.ipynb).

**Comment by sachinprasadhs** [comment]
It is due to the default epsilon value in the `l2 normalize`, basically `x_inv_norm` will be having the value of in the return statement here https://github.com/tensorflow/tensorflow/blob/v2.9.1/tensorflow/python/ops/nn_impl.py#L644-L700.
Below is the debug details for the same which shows the details
<img width="1244" alt="image" src="https://user-images.githubusercontent.com/73069040/184025073-00ddc4a6-6a83-4034-b794-dd08cbadb69d.png">

**Comment by VictoriaGriffith** [comment]
Thanks @sachinprasadhs for pointing this out! If this is an expected behavior, I will close this issue.

**Comment by VictoriaGriffith** [comment]
hi @sachinprasadhs , any update on this issue?

**Comment by cantonios** [comment]
The function `x/norm(x)` is not differentiable at 0, so what do you expect it to be?  Without `epsilon`, a straight-forward implementation would produce `NaN`.  With `epsilon`, the piecewise definition of the function is
```
l2_normalize(x) =  x/sqrt(epsilon),    if x < sqrt(epsilon)
                                x/norm(x),            otherwise
```
so for small x, the gradient of this is `1/sqrt(epsilon)`, which is exactly what is returned.

**Comment by VictoriaGriffith** [comment]
Thanks @cantonios for the detailed answer, that hakes sense and I will close this issue.
